### PR TITLE
feat(collections): adds takeLastWhile

### PR DESCRIPTION
**Pull request by grian32**
Not sure if it's possible to reverse the original array/get the elements going from the end while not copying the array so it doesn't mutate.

### CODE REVIEW

**Comment by LionC** [comment]
I dont think you should copy **and** reverse the whole array just to use the same imolementation `takeWhile` uses. That takes up more memory than needed.

I would just do a simple for loop starting from the end counting back, pushing until it hits a non-matching element as a baseline and go from there.

**Comment by grian32** [comment]
> I dont think you should copy **and** reverse the whole array just to use the same imolementation `takeWhile` uses. That takes up more memory than needed.
> 
> I would just do a simple for loop starting from the end counting back, pushing until it hits a non-matching element as a baseline and go from there.

I mostly tried to use the same impl as `takeWhile` because it was pretty performant, I'll see what I can do though.

**Comment by grian32** [comment]
I was able to to do this without copying and reversing the original array @LionC , I needed to use a more naive implementation though. Surprisingly it did not affect performance. Ready for review I think.

**Comment by LionC** [comment]
Looking at the implementation now with not-as-tired eyes, I think this might actually be faster if you use `collections/findLastIndex` and then `slice`, as it will allocate the new array only once instead of reallocating on every push.

Sorry for the back and forth here :-)

**Comment by grian32** [comment]
> Looking at the implementation now with not-as-tired eyes, I think this might actually be faster if you use `collections/findLastIndex` and then `slice`, as it will allocate the new array only once instead of reallocating on every push.

I don't think I fully understand what you mean but by toying around with an implementation using `collections/findLastIndex`, what tests I can get to pass(which is most of them) don't really have a visible perf improvement.

**Comment by LionC** [comment]
> > Looking at the implementation now with not-as-tired eyes, I think this might actually be faster if you use `collections/findLastIndex` and then `slice`, as it will allocate the new array only once instead of reallocating on every push.
> 
> I don't think I fully understand what you mean but by toying around with an implementation using `collections/findLastIndex`, what tests I can get to pass(which is most of them) don't really have a visible perf difference(other than maybe being 1-2ms slower).

What I meant is that the current implementation `.push()`es into the array, causing it to reallocate several times to make space for new elements. It also needs to iterate the whole array by calling `.reverse()` in the end.

If you use `findLastIndex()`, you can leverage `.slice()` instead, which is optimized by v8 and can pretty much just copy the memory directly once. But that is obviously a question of benchmarks in the end :-)

**Comment by grian32** [comment]
> What I meant is that the current implementation `.push()`es into the array, causing it to reallocate several times to make space for new elements. It also needs to iterate the whole array by calling `.reverse()` in the end.
> 
> If you use `findLastIndex()`, you can leverage `.slice()` instead, which is optimized by v8 and can pretty much just copy the memory directly once. But that is obviously a question of benchmarks in the end :-)

Yeah so you're saying to use `findLastIndex()` to get the indexes to slice at, I tried to do that and could get a somewhat working implementation although it failed horribly at returning empty array when predicate matched nothing/same array when predicate matched everything

File: collections/take_last_while.ts [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+
+  for (let i = length - 1; i >= 0; i--) {
+    if (predicate(array[i])) {
+      newArray.push(array[i]);
+    } else {
+      break;
+    }
+  }
+
+  return newArray.reverse();
```
Comment (evanwashere) [inline collections/take_last_while.ts:33]:
```suggestion
  let offset = array.length;
  while (0 < offset && predicate(array[offset - 1])) offset--;

  return array.slice(offset, array.length);
```

**Review by evanwashere — CHANGES_REQUESTED** [review]

File: collections/take_last_while.ts (hunk 1 above)
Comment (grian32) [reply collections/take_last_while.ts:33]: Thanks for this!

Comment (LionC) [reply collections/take_last_while.ts]: Isn't

```ts
let offset = array.length;
while (0 < offset && predicate(array[offset - 1])) offset--;
```

pretty much `findLastIndex()`, which we already have implemented in the module?

Comment (LionC) [reply collections/take_last_while.ts]: Update:

Doing some spontanous benchmarking, I can not find any significant stable performance difference between

```ts
  const index = findLastIndex(array, (it) => !predicate(it))

  if (index === undefined) {
      return []
  }

  return array.slice(index + 1);
```

and

```ts
  let offset = array.length;
  while (0 < offset && predicate(array[offset - 1])) offset--;

  return array.slice(offset, array.length);
```

I guess either is fine, I just want to emphasize that if we can DRY code up without a significant impact, we probably should for maintenance reasons.

File: collections/README.md [hunk 2]
```diff
@@ -527,6 +527,25 @@ assertEquals(
 );
 ```
 
+### takeLastWhile
+
+Returns all elements in the given collection until the first element that does
+not match the given predicate.
```
Comment (kt3k) [inline collections/README.md:533]: I think we need to mention that this function collect elements 'from the back'

File: collections/README.md [hunk 3]
```diff
@@ -527,6 +527,25 @@ assertEquals(
 );
 ```
 
+### takeLastWhile
+
+Starting from the last element, returns all elements in the given collection
```
Comment (LionC) [inline collections/README.md:532]: This does sound like the order would be reversed, but it actually isnt

File: collections/take_last_while.ts [hunk 4]
```diff
@@ -0,0 +1,27 @@
+// Copyright 2018-2021 the Deno authors. All rights reserved. MIT license.
+/**
+ * Starting from the last element, returns all elements in the given collection
+ * until the first element that does not match the given predicate.
```
Comment (LionC) [inline collections/take_last_while.ts:4]: Same here obv

**Review by LionC — CHANGES_REQUESTED** [review]

File: collections/README.md (hunk 3 above)
Comment (grian32) [reply collections/README.md:532]: What about `Returns all elements in the given collection until the last element that does not match the given predicate.`, its the equivalent of what I have in the `dropLastWhile` PR, @kt3k

Comment (LionC) [reply collections/README.md:532]: But that is not correct - it returns all element in the given array (we should not use the word collection here) **after** the last element that does not match the given predicate.

But I dont know if that is the most elegant way to phrase it.

Comment (grian32) [reply collections/README.md:532]: I don't really have any better ideas so I may just go with `Returns all elements in the given array after the last element that does not match the given predicate.`

File: collections/README.md [hunk 5]
```diff
@@ -512,8 +512,8 @@ console.assert(
 
 ### takeWhile
 
-Returns all elements in the given collection until the first element that does
-not match the given predicate.
+Returns all elements in the given array after the last element that does not
```
[the remark this answers is not in the document]
Comment (grian32) [reply collections/README.md]: Yeah looks like it, I think I've reverted it, what I get for commiting stuff without checking

**Review by LionC — APPROVED** [review]
LGTM, pinging @kt3k

**Review by kt3k — APPROVED** [review]
@grian32 LGTM. Thanks!

**Pull request merged**